docs/javascript-unit-tests.md

```diff
@@ -0,0 +1,712 @@
+    1. [Testing user interaction](#testing-user-interaction)
+1. [Running unit tests](#running-unit-tests)
+
+
+
+
+## Test-driven development
+
+We recommend using test-driven development (TDD) when coding new JavaScript.
+The general concept is to start by writing your test **first**,
+with the expected behavior and functionality well-described,
+**then** you write the code that makes the test pass.
+[A good pithy summary](http://www.javiersaldana.com/tech/2014/11/26/refactoring-the-three-laws-of-tdd.html)
+is:
+
+> 1.  Write only enough of a unit test to fail.
+> 2.  Write only enough production code to make the failing unit test pass.
+
+Then repeat that process until you have written all of the code you need.
+
+[Read this primer on test-driven development](https://medium.freecodecamp.org/test-driven-development-what-it-is-and-what-it-is-not-41fa6bca02a2?gi=3c8d8b476cc9)
+to learn more about how it differs from
+the typical approach to programming and unit tests.
+
+
+
+
+## Setting up tests
+
+
+### New test file from sample
+
+For this guide, we’ll use very basic sample code files
+to illustrate how to use the test framework in cfgov-refresh
+and how to test very common code patterns.
+
+Another common approach is to look for existing tests
+that are testing something similar to what you are writing now.
+Feel free to do so and substitute another test file for
+the `sample-spec.js` file referenced below.
+For links to existing tests,
+[see the “Running tests” section](#running-tests) on this page.
+
+Now, let’s begin! Let’s make a new unit test fail,
+then we will make it pass, following the principles of TDD.
+
+1. Create 2 new files: a test file, and a JavaScript file.
+   Their names should match, with the test file adding a `-spec` suffix.
+   For example, if your script is named `my-javascript.js`,
+   then your test is named `my-javascript-spec.js`.
+1. Copy the code from this
+   [`sample-spec.js` file](https://github.com/cfpb/cfgov-refresh/blob/fc9ecc002f933a432322bd15f54be5eb2cffa6e3/test/unit_tests/js/modules/sample-spec.js)
+   into your own test file.
+1. Copy and paste from
+   [`sample.js`](https://github.com/cfpb/cfgov-refresh/blob/fc9ecc002f933a432322bd15f54be5eb2cffa6e3/cfgov/unprocessed/js/modules/sample.js)
+   into your new JavaScript file.
+
+
+### Folder structure (where to put your JavaScript and tests)
+
+JavaScript unit test files belong in the
+[`test/unit_tests/`](https://github.com/cfpb/cfgov-refresh/tree/master/test/unit_tests)
+directory.
+
+The folder structure of the test files mirrors
+the structure of the project JavaScript in
+[`cfgov/unprocessed/js/`](https://github.com/cfpb/cfgov-refresh/tree/master/cfgov/unprocessed/js).
+
+So my `sample-spec.js` file is in `test/unit_tests/modules/`
+and my `sample.js` file is in `cfgov/unprocessed/js/modules/`.
+I put these files in the `modules` folder because
+they are not atomic components, they are just sample files.
```

Comment:
For someone unfamiliar with atomic components, the wording here is a little bit confusing until reading the [atomic structure](https://cfpb.github.io/cfgov-refresh/atomic-structure/) doc. What about changing it slightly to something like:

> When considering exactly where to place JavaScript in these directories, it might be helpful to review the documentation about [atomic components in cfgov-refresh](../atomic-structure). JavaScript corresponding to atomic elements should go into the appropriate subfolder for the type of element being implemented. In our case, `sample.js` and `sample-spec.js` don't relate to atomic elements, so they can be placed into the uncategorized `modules` subfolders: `cfgov/unprocessed/js/modules` and `test/unit_tests/modules`, respectively.